serde_json_path_core/src/spec/integer.rs

```diff
@@ -0,0 +1,145 @@
+//! Representation of itegers in the JSONPath specification
+//!
+//! The JSONPath specification defines some rules for integers used in query strings (see [here][spec]).
+//!
+//! [spec]: https://www.rfc-editor.org/rfc/rfc9535.html#name-overview
+
+use std::{
+    num::{ParseIntError, TryFromIntError},
+    str::FromStr,
+};
+
+/// An integer for internet JSON ([RFC7493][ijson])
+///
+/// The value must be within the range [-(2<sup>53</sup>)+1, (2<sup>53</sup>)-1]).
+///
+/// [ijson]: https://www.rfc-editor.org/rfc/rfc7493#section-2.2
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct Integer(i64);
+
+/// The maximum allowed value, 2^53 - 1
+const MAX: i64 = 9_007_199_254_740_992 - 1;
+/// The minimum allowed value (-2^53) + 1
+const MIN: i64 = -9_007_199_254_740_992 + 1;
+
+impl Integer {
+    fn try_new(value: i64) -> Result<Self, IntegerError> {
+        if !(MIN..=MAX).contains(&value) {
+            Err(IntegerError::OutOfBounds)
+        } else {
+            Ok(Self(value))
+        }
+    }
+
+    fn check(&self) -> bool {
+        (MIN..=MAX).contains(&self.0)
+    }
+
+    /// Produce an [`Integer`] with the value 0
+    pub fn zero() -> Self {
+        Self(0)
+    }
+
+    /// Get an [`Integer`] from an `i64`
+    ///
+    /// This is intended for initializing an integer with small, non-zero numbers.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if the inputted value is out of the valid range
+    /// [-(2<sup>53</sup>)+1, (2<sup>53</sup>)-1]).
+    pub fn from_i64_unchecked(value: i64) -> Self {
+        Self::try_new(value).expect("value is out of the valid range")
+    }
+
+    /// Take the absolute value, producing `None` if the resulting value is outside
+    /// the valid range [-(2<sup>53</sup>)+1, (2<sup>53</sup>)-1]).
+    pub fn checked_abs(mut self) -> Option<Self> {
+        self.0 = self.0.checked_abs()?;
+        self.check().then_some(self)
+    }
+
+    /// Add the two values, producing `None` if the resulting value is outside the
+    /// valid range [-(2<sup>53</sup>)+1, (2<sup>53</sup>)-1]).
+    pub fn checked_add(mut self, rhs: Self) -> Option<Self> {
+        self.0 = self.0.checked_add(rhs.0)?;
+        self.check().then_some(self)
+    }
```

Comment:
Out of curiosity (applies to all the mutating methods): Is this pattern of both mutating `self` and returning `self` that common? It might be a bit error-prone, e.g. users might not expect (or notice) that `self` is mutated. Should instead a new `Integer` be returned, and `self` not be mutated?

Also, I am not sure if this approach of first doing `self.0 = ...` and then calling `self.check()` is safe. If the caller ignores the result, or continues to use the `Integer` afterwards, it now contains an invalid value.

Reply:
Yeah, this is definitely a foot gun. I don't think it is a problem from where `Integer` is used in `serde_json_path`, but I will definitely change this.

Since these changes are not released yet I can make breaking changes to the API of `Integer`.

Reply:
Ah sorry, my comment was wrong. Since it is taking `mut self` and not `&mut self` it was actually behaving as expected since `Integer` is `Copy`, and actually ended up creating a copy. But the code might still be a bit confusing then.